File: Dev/unitTest_Engine_cs/BugFix/RepeatedlyAddAndRemoveObject.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using asd;

namespace unitTest_Engine_cs.BugFix
{
	class MyObject_RepeatedlyAddAndRemoveObject : TextureObject2D
	{
		protected override void OnAdded()
		{
			Console.WriteLine("OnAdded");
			Assert.AreNotEqual(null, Layer);
		}

		protected override void OnRemoved()
		{
			Console.WriteLine("OnRemoved");
		}
	}

	class RepeatedlyAddAndRemoveObject : EngineTest
	{
		public RepeatedlyAddAndRemoveObject()
			: base(60)
		{
		}

		protected override void OnStart()
		{
			var scene = new Scene();
			var layer = new Layer2D();
			var layer2 = new Layer2D();
			var obj = new MyObject_RepeatedlyAddAndRemoveObject();

			Engine.ChangeScene(scene);
			scene.AddLayer(layer);

			for (int i = 0; i < 4; i++)
			{
				layer.AddObject(obj);
				layer.RemoveObject(obj);
			}
			for(int i = 0; i < 4; i++)
			{
				layer2.AddObject(obj);
				layer2.RemoveObject(obj);
			}
		}
	}
}

File: Dev/unitTest_Engine_cpp_gtest/BugFix/RepeatedlyAddAndRemoveObject.cpp
#include <Altseed.h>
#include <gtest/gtest.h>
#include <memory>
#include "../EngineTest.h"

using namespace std;
using namespace asd;

class MyClass_RepeatedlyAddAndRemoveObject : public TextureObject2D
{
protected:
	void OnAdded() override
	{
		cout << "OnAdded" << endl;
		ASSERT_NE(nullptr, GetLayer());
	}

	void OnRemoved() override
	{
		cout << "OnRemoved" << endl;
	}
};

class BugFix_RepeatedlyAddAndRemoveObject : public EngineTest
{
private:
	Layer2D::Ptr layer;
	shared_ptr<MyClass_RepeatedlyAddAndRemoveObject> obj;

public:
	BugFix_RepeatedlyAddAndRemoveObject(bool isOpenGLMode)
		: EngineTest(asd::ToAString("RepeatedlyAddAndRemoveObject"), isOpenGLMode, 60)
	{
	}

protected:
	void OnStart()
	{
		auto scene = make_shared<Scene>();
		layer = make_shared<Layer2D>();
		obj = make_shared<MyClass_RepeatedlyAddAndRemoveObject>();
	}

	void OnUpdating()
	{
		if (m_currentTime == 20)
		{
			for (size_t i = 0; i < 3; i++)
			{
				layer->AddObject(obj);
				layer->RemoveObject(obj);
			}
		}
	}

	void OnUpdated()
	{
	}

	void OnFinish()
	{
	}
};

ENGINE_TEST(BugFix, RepeatedlyAddAndRemoveObject)